### ISSUE TITLE
How to get curent metric within Horizontal Pod Autoscaler

### ISSUE THREAD

**Issue by akash123srm** [kubernetes/kube-state-metrics#2143]
**How to get curent metric within Horizontal Pod Autoscaler on Grafana Dashboards**:

**Why is this needed**:

**Hi guys, is there a way within kube-state-metrics for HPA that would fetch the current metric for CPU utilization.
I now there is one availaible for target metric i.e. kube_horizontalpodautoscaler_status_target_metric. Is there something similar that could be used to fetch the actual/current metiric**

**Additional context**

**Comment by buger** [comment]
I'm working on some AI architect project, and if you do not mind I will do a PoC on your project, by helping answering some queries. If you want me to stop just let me know. All the answers are based fully AI generated:

Based on the description provided, it does not look like kube-state-metrics exposes a metric for the current CPU utilization that a HorizontalPodAutoscaler is scaling on.

The `kube_horizontalpodautoscaler_status_current_metric` metric provides the current metric values for each target metric the HPA is tracking, but this metric was removed in kube-state-metrics v2.0.0 according to the [changelog](https://github.com/kubernetes/kube-state-metrics/blob/master/CHANGELOG.md#v200--2021-01-13).

Some options to get the current CPU utilization metric for HPA scaling:

-   The HPA controller writes the current metric values to the `status.currentMetrics` field of the HorizontalPodAutoscaler object. You can query this field directly via the Kubernetes API, e.g. using kubectl.
-   The Kubernetes metrics server component exposes the current CPU utilization for pods and nodes via the `pods/metrics_cpu_usage` and `nodes/metrics_cpu_usage` APIs. Prometheus can scrape these endpoints directly to record the current CPU usage.
-   You could run a Prometheus query against the raw Kubernetes metrics to calculate the current CPU usage percentage for the pods selected by the HPA. For example:

`sum(rate(container_cpu_usage_seconds_total{namespace="<namespace>", pod=~"<hpa-selector-labels>"}[5m])) /
sum(kube_pod_container_resource_requests_cpu_cores{namespace="<namespace>", pod=~"<hpa-selector-labels>"}) * 100`

So in summary, while kube-state-metrics itself doesn't have this metric, there are ways to get the current CPU utilization either directly from the Kubernetes API or by querying the raw metrics exposed by the metrics server. Let me know if this helps explain the options!

**Comment by akash123srm** [comment]
Hi Leonid,
Thanks a lot mate for your suggestion.
Yes, as  suggested, I ran the prometheus query against raw Kubernetes
Metrics and was able to get the CPU utilization and it worked.

Thanks once again. Let me know if I can be of any help.

Regards

**Comment by dgrisonnet** [comment]
As suggested before, you should really on the `container_cpu_usage_seconds_total` metric from kubelet in order to get the CPU utilization of your containers.
